File: TheaterSuggestions/CSharp/TheaterSuggestions.Tests/SeatAllocator.cs
using System.Collections.Generic;

namespace SeatsSuggestions.Tests
{
    public class SeatAllocator
    {
        private readonly AuditoriumSeatingAdapter _auditoriumSeatingAdapter;

        public SeatAllocator(AuditoriumSeatingAdapter auditoriumSeatingAdapter)
        {
            _auditoriumSeatingAdapter = auditoriumSeatingAdapter;
        }

        public SuggestionsMade MakeSuggestions(string showId, int partyRequested)
        {
            var auditoriumSeating = _auditoriumSeatingAdapter.GetAuditoriumSeating(showId);

            var suggestionsMade = new SuggestionsMade(showId, partyRequested);

            var numberOfSuggestions = 3;

            suggestionsMade.Add(GiveMeSeveralSuggestionFor(partyRequested, auditoriumSeating, numberOfSuggestions,
                PricingCategory.First));
            suggestionsMade.Add(GiveMeSeveralSuggestionFor(partyRequested, auditoriumSeating, numberOfSuggestions,
                PricingCategory.Second));
            suggestionsMade.Add(GiveMeSeveralSuggestionFor(partyRequested, auditoriumSeating, numberOfSuggestions,
                PricingCategory.Third));
            suggestionsMade.Add(GiveMeSeveralSuggestionFor(partyRequested, auditoriumSeating, numberOfSuggestions,
                PricingCategory.Mixed));

            if (!suggestionsMade.MatchExpectations())
            {
                return new SuggestionNotAvailable(showId, partyRequested);
            }

            return suggestionsMade;
        }

        private static IReadOnlyCollection<SuggestionMade> GiveMeSeveralSuggestionFor(int partyRequested,
            AuditoriumSeating auditoriumSeating, int numberOfSuggestions, PricingCategory pricingCategory)
        {
            var foundedSuggestions = new List<SuggestionMade>();
            for (var i = 0; i < numberOfSuggestions; i++)
            {
                foundedSuggestions.Add(auditoriumSeating.MakeSuggestionFor(partyRequested, pricingCategory));
            }

            return foundedSuggestions;
        }
    }
}

File: TheaterSuggestions/Java/SeatsSuggestionsTests/src/test/java/com/baasie/SeatsSuggestionsTests/SeatAllocator.java
package com.baasie.SeatsSuggestionsTests;

import com.google.common.collect.ImmutableList;

import java.util.ArrayList;
import java.util.List;

public class SeatAllocator {

    private final AuditoriumSeatingAdapter auditoriumSeatingAdapter;

    public SeatAllocator(AuditoriumSeatingAdapter auditoriumLayoutAdapter) {
        this.auditoriumSeatingAdapter = auditoriumLayoutAdapter;
    }

    public SuggestionsMade makeSuggestion(String showId, int partyRequested) {

        AuditoriumSeating auditoriumSeating = auditoriumSeatingAdapter.getAuditoriumSeating(showId);

        SuggestionsMade suggestionsMade = new SuggestionsMade(showId, partyRequested);

        int numberOfSuggestions = 3;

        suggestionsMade.add(giveMeSeveralSuggestionFor(partyRequested, auditoriumSeating, numberOfSuggestions,
                PricingCategory.First));
        suggestionsMade.add(giveMeSeveralSuggestionFor(partyRequested, auditoriumSeating, numberOfSuggestions,
                PricingCategory.Second));
        suggestionsMade.add(giveMeSeveralSuggestionFor(partyRequested, auditoriumSeating, numberOfSuggestions,
                PricingCategory.Third));
        suggestionsMade.add(giveMeSeveralSuggestionFor(partyRequested, auditoriumSeating, numberOfSuggestions,
                PricingCategory.Mixed));

        if (!suggestionsMade.matchExpectations()) {
            return new SuggestionNotAvailable(showId, partyRequested);
        }

        return suggestionsMade;
    }

    private static ImmutableList<SuggestionMade> giveMeSeveralSuggestionFor(int partyRequested,
                                                                            AuditoriumSeating auditoriumSeating, int numberOfSuggestions, PricingCategory pricingCategory) {
        List<SuggestionMade> foundedSuggestions = new ArrayList<>();
        for (int i = 0; i < numberOfSuggestions; i++) {
            foundedSuggestions.add(auditoriumSeating.makeSuggestionFor(partyRequested, pricingCategory));
        }

        return ImmutableList.copyOf(foundedSuggestions);
    }
}
